Mobx-6 migration

ixrock opened the pull request:
Attempt to upgrade from 5.x to 6.x. 

(re)based on https://github.com/lensapp/lens/pull/2520

**Main pitfalls:**
- when using `mobx` decorators (e.g `@observable`) class constructor must include `makeObservable(this)`
-  don't use`mobx.toJS()` with partial observable data, e.g. `toJS({x: 1, y: observable([])})` still returns `result.y` as `observable` => use `common/utils/toJS` to handle such cases.
- don't apply `@action` for overriding class methods or use extra `@override` from `mobx` package
- don't use `@observable static` for classes (there is no way to wrap it with `makeObservable`)
- second argument in `mobx.toJS` has removed, e.g. `toJS(data, {recursiveEverything: true})` will throw an error.
- don't cache `when(() => this.someObservableBoolean)` in class fields (might not work as you expected)

**Try it locally:**
- `git checkout mobx6-migration && yarn`
- `make clean && make dev`

**More info:**
https://mobx.js.org/migrating-from-4-or-5.html

Fixes #1536

ixrock committed:
mobx-6 migration -- part 1

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
mobx-6 migration -- part 2 (npx mobx-undecorate --keepDecorators)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
mobx-6 migration -- part 3 (more fixes)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
unwrap possible observables from IPC-messaging

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
mobx-6 migration -- remove @autobind as class-decorator

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
mobx-6: replacing @autobind() as method-decorator to @boundMethod

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
mobx-6: use toJS()-wrapper since monkey-patching require(mobx).toJS doesn't work

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
removed `@observable static`

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
use {useDefineForClassFields: true} in tsconfig.json

Signed-off-by: Roman <ixrock@gmail.com>

jakolehm committed:
remove ExtendedObservableMap

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

ixrock committed:
fix: removed makeObservable(this) from "terminal-tab.tsx"

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/renderer/kube-object.store.ts

ixrock committed:
storage-helper refactoring

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
normalize usages of #observable-value.toJSON() / attempt to catch the wind

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
refactoring, more possible branch fixes + lint

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
debugging cluster-view error -- part 1

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: refreshing cluster-view on ready

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: various app-crashes related to KubeObject.spec.* access from "undefined"
fix: config-map-details crash

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: namespace-store refactoring / saving selected-namespaces to external json-file

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: don't cache mobx.when(() => this.someObservable) cause might not work as expected due later call of makeObservable(this) in constructor

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: app-crash on editing k8s resource

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: restore "all namespaces" on page reload

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
- fix: persist table-sort params and cluster-view's sidebar state to lens-local-storage
- new-feature: auto-open main-window's devtools in development-mode (yes/no/ugly?)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: crd definition details -> crashing with <AceEditor mode="json"> (added missing mode-file in ace-editor.tsx)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: crd definitions -> groups selector couldn't deselect last selected option

Signed-off-by: Roman <ixrock@gmail.com>

ixrock commented on docs/extensions/guides/stores.md:
```diff
@@ -42,18 +42,17 @@ export class ExamplePreferencesStore extends Store.ExtensionStore<ExamplePrefere
   }
 
   toJSON(): ExamplePreferencesModel {
-    return toJS({
```
Wrapping with `toJS()` probably not necessary since:

- properly handled in `BaseStore.enableSync()` for reacting to every observable from `BaseStore.toJSON()`
- sanitizing payload for IPC-messaging (removing all possible observables for `handleRequest`, `requestMain` and `broadcastMessage` in `ipc.tsx`)

ixrock commented on src/common/base-store.ts:
```diff
@@ -20,13 +19,18 @@ export interface BaseStoreParams<T = any> extends ConfOptions<T> {
   @observable isLoaded = false;
 
+  get whenLoaded() {
+    return when(() => this.isLoaded);
```
Seems like `when` evaluates callback to detect observables before `makeObservable(this)` is done in `constructor`.

Nokel81 replied:
Why do you think this? That getter shouldn't be called until after that `makeObservable(this)` executes.

ixrock replied:
I was talking about how it was done before.

Nokel81 replied:
Oh okay makes sense.

ixrock commented on src/common/base-store.ts:
```diff
@@ -93,7 +97,11 @@ export abstract class BaseStore<T = any> extends Singleton {
     this.syncDisposers.push(
-      reaction(() => this.toJSON(), model => this.onModelChange(model), this.params.syncOptions),
+      reaction(
+        () => toJS(this.toJSON()), // unwrap possible observables and react to everything
```
This probably should have better compatibility with existing extensions after migration.

ixrock commented on src/extensions/extensions-store.ts:
```diff
@@ -47,9 +49,7 @@ export class ExtensionsStore extends BaseStore<LensExtensionsStoreModel> {
-      extensions: this.state.toJSON(),
-    }, {
-      recurseEverything: true
+      extensions: Object.fromEntries(this.state),
```
FYI: `Object.fromEntries` might accept `Map` or `ObservableMap`

Nokel81 replied:
Might accept? I think it does, since what it wants is an `Iterable<[string, T]>`

ixrock commented on src/main/cluster-manager.ts:
```diff
@@ -137,9 +139,7 @@ export class ClusterManager extends Singleton {
 
 export function catalogEntityFromCluster(cluster: Cluster) {
-  return new KubernetesCluster(toJS({
-    apiVersion: "entity.k8slens.dev/v1alpha1",
-    kind: "KubernetesCluster",
```
Those are already defined as `public readonly` in `KubernetesCluster`

ixrock commented on src/renderer/api/endpoints/cluster.api.ts:
```diff
@@ -54,7 +54,7 @@ export class Cluster extends KubeObject {
   static apiBase = "/apis/cluster.k8s.io/v1alpha1/clusters";
 
-  spec: {
+  declare spec: {
```
This is required with `"useDefineForClassFields": true` in `tsconfig.ts` when field already defined at parent's class.

ixrock commented on src/renderer/components/+apps/apps.tsx:
```diff
@@ -3,12 +3,10 @@ import { observer } from "mobx-react";
 @observer
 export class Apps extends React.Component {
   static get tabRoutes(): TabLayoutRoute[] {
-    const query = namespaceUrlParam.toObjectParam();
```
Keeping selected namespaces in URL not necessary anymore since `lens-local-storage` persists state in external json file.

Nokel81 replied:
Thanks for finding this. I was thinking that was the case too.

ixrock commented on src/renderer/components/+config-maps/config-map-details.tsx:
```diff
@@ -39,7 +44,10 @@ export class ConfigMapDetails extends React.Component<Props> {
 
     try {
       this.isSaving = true;
-      await configMapsStore.update(configMap, { ...configMap, data: this.data.toJSON() });
```
N.B. In _mobx@6_ `ObservableMap.toJSON()` returns array of entries, instead of plain-object.

ixrock commented on src/renderer/components/+namespaces/namespace.store.ts:
```diff
@@ -130,41 +100,50 @@ export class NamespaceStore extends KubeObjectStore<Namespace> {
   }
 
   @action
-  setContext(namespace: string | string[]) {
```
There are some breaking changes with names of class methods. 
If this is critical we could add wrappers with old names and mark as `@deprecated`. Wdyt?

Nokel81 replied:
I don't think you should bother. I have been working (slowly and around other things) to do an almost complete refactor of how we deal with namespaces. So I don't think we should bother trying to support the older (and frankly very complex) form.

ixrock commented on src/renderer/components/cluster-manager/cluster-view.tsx:
```diff
@@ -1,81 +1,70 @@
-          navigate(catalogURL());
-        }
-      })
-    ]);
```
Not sure if this is needed anymore. Wdyt? cc @jakolehm

Nokel81 replied:
Doesn't that redirect you to the catalog if you disconnect from the currently viewed cluster?

ixrock replied:
I have to check this more.. I wasn't able to disconnect from cluster:
<img width="283" alt="Screenshot 2021-05-11 at 16 24 55" src="https://user-images.githubusercontent.com/6377066/117822799-83f94000-b275-11eb-9ebc-89c73517696d.png">

Maybe because of this weirdness (went to catalog from left-bottom icon). It show stale state for some reasons?
<img width="870" alt="Screenshot 2021-05-11 at 16 25 11" src="https://user-images.githubusercontent.com/6377066/117822807-8491d680-b275-11eb-93de-e7f62563407b.png">

Nokel81 replied:
That is weird because that should be an option.

ixrock commented on src/renderer/utils/storageHelper.ts:
```diff
@@ -1,14 +1,10 @@
 import { isEqual, isFunction, isPlainObject } from "lodash";
 import logger from "../../main/logger";
 
-setAutoFreeze(false); // allow to merge observables
-enableMapSet(); // allow merging maps and sets
```
Moved to `src/common/configure-packages.ts` + mobx configurations.

aleksfront commented on src/common/search-store.ts:
```diff
@@ -107,12 +108,12 @@ export class SearchStore {
   }
 
-  @autobind()
+  @boundMethod
```
Since it will be located in many different places is it possible to shorten name to just `@bound`?

Nokel81 replied:
Is there a good reason to not just use the `field = () => { ... };`

ixrock replied:
@Nokel81 for example for when overriding a method in sub-class + that way it keeps single method definition in prototype (less memory consumption in some cases)

ixrock replied:
@aleksfront not sure if a short name is better (I hope you use auto-import / auto-complete in your IDE:)

aleksfront commented on src/main/window-manager.ts:
```diff
@@ -84,6 +85,7 @@ export class WindowManager extends Singleton {
           shell.openExternal(url);
         })
         .on("dom-ready", () => {
+          this.mainWindow.webContents.openDevTools({ mode: "right", activate: isDevelopment });
```
Very convenient!

ixrock replied:
Probably better to move this to separated PR.

ixrock committed:
refactoring: extensions-api exports clarification for "@k8slens/extensions"

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: various app-crashes related to kube-events (events page, some details page, overview, etc.)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Reverted "use {useDefineForClassFields: true} in tsconfig.json" (various app-crash fixes)
This flag seems to be not possible to use with class-inheritance in some cases.

Example / demo:
`KubeObject` class has initial type definitions for the fields like: "metadata", "kind", etc.
and constructor() has Object.assign(this, data);
Meanwhile child class, e.g. KubeEvent inherited from KubeObject and has it's own extra type definitions for underlying resource, e.g. "involvedObject", "source", etc.

So calling super(data) doesn't work as expected for child class as it's own type definitions overwrites data from parent's constructor with `undefined` at later point.

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/renderer/components/+user-management/user-management.tsx
#	src/renderer/components/layout/sidebar.tsx
#	yarn.lock

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/common/catalog/catalog-category-registry.ts
#	src/common/catalog/catalog-entity-registry.ts
#	src/common/hotbar-store.ts
#	src/common/search-store.ts
#	src/common/utils/autobind.ts
#	src/common/utils/extended-map.ts
#	src/common/utils/toggle-set.ts
#	src/extensions/extension-api.ts
#	src/extensions/renderer-api/navigation.ts
#	src/main/catalog-pusher.ts
#	src/main/catalog-sources/kubeconfig-sync.ts
#	src/renderer/api/catalog-entity-registry.ts
#	src/renderer/api/endpoints/cluster-role.api.ts
#	src/renderer/api/endpoints/endpoint.api.ts
#	src/renderer/api/endpoints/poddisruptionbudget.api.ts
#	src/renderer/api/endpoints/podsecuritypolicy.api.ts
#	src/renderer/api/endpoints/role-binding.api.ts
#	src/renderer/api/endpoints/service-accounts.api.ts
#	src/renderer/api/endpoints/service.api.ts
#	src/renderer/api/endpoints/storage-class.api.ts
#	src/renderer/api/websocket-api.ts
#	src/renderer/components/+catalog/catalog-entity.store.ts
#	src/renderer/components/+cluster/cluster-overview.store.ts
#	src/renderer/components/+config-autoscalers/hpa.store.ts
#	src/renderer/components/+config-limit-ranges/limit-ranges.store.ts
#	src/renderer/components/+custom-resources/crd-resource.store.ts
#	src/renderer/components/+custom-resources/crd.store.ts
#	src/renderer/components/+namespaces/namespace.store.ts
#	src/renderer/components/+network-ingresses/ingress.store.ts
#	src/renderer/components/+storage-volume-claims/volume-claim.store.ts
#	src/renderer/components/+user-management-service-accounts/service-accounts.store.ts
#	src/renderer/components/+workloads-daemonsets/daemonsets.store.ts
#	src/renderer/components/+workloads-deployments/deployments.store.ts
#	src/renderer/components/+workloads-replicasets/replicasets.store.ts
#	src/renderer/components/+workloads-statefulsets/statefulset.store.ts
#	src/renderer/components/command-palette/command-dialog.tsx
#	src/renderer/components/dock/dock-tab.store.ts
#	src/renderer/components/dock/edit-resource.store.ts
#	src/renderer/components/dock/install-chart.store.ts
#	src/renderer/components/dock/log.store.ts
#	src/renderer/components/dock/upgrade-chart.store.ts
#	src/renderer/components/item-object-list/page-filters.store.ts
#	src/renderer/navigation/page-param.ts
#	src/renderer/theme.store.ts

ixrock committed:
master-merge lint-fixes

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/renderer/components/hotbar/hotbar-icon.tsx

ixrock committed:
catalog.tsx / catalog-entities.store.ts refactoring & fixes

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: Catalog -> Browse all tab

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: CommandPalette doesn't appear from global menu by click/hotkey

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
- Merging interfaces & classses to avoid overwriting fields from parent's super(data)-call with Object.assign(this, data). Otherwise use "declare" keyword at class field definition.

- Revamping {useDefineForClassFields: true} to avoid issues with non-observable class fields in some cases (from previous commit):

```
@observer
export class CommandContainer extends React.Component<CommandContainerProps> {
  // without some defined initial value "commandComponent" is non-observable for some reasons
  // when tsconfig.ts has {useDefineForClassFields:false}
  @observable.ref commandComponent: React.ReactNode = null;

  constructor(props: CommandContainerProps) {
    super(props);
    makeObservable(this);
  }
```

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
update KubeObject class type definition

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/extensions/extension-loader.ts:
```diff
@@ -332,13 +341,10 @@ export class ExtensionLoader extends Singleton {
 
-  broadcastExtensions(main = true) {
-    broadcastMessage(main ? ExtensionLoader.extensionsMainChannel : ExtensionLoader.extensionsRendererChannel, Array.from(this.toJSON()));
+  broadcastExtensions(channel = ExtensionLoader.extensionsMainChannel) {
```
This shouldn't be an optional parameter.

ixrock replied:
Fixed

Nokel81 commented on src/common/catalog/catalog-entity-registry.ts:
```diff
@@ -19,27 +19,33 @@
-    return Array.from(iter.flatMap(this.sources.values(), source => source.get()));
+    return Array.from(this.sources.values())
+      .map(source => source.get())
+      .flat();
```
Please revert things like this. This just clutters up the diff making it hard to review.

ixrock replied:
I wasn't sure previous code was working with mobx-6. So, I'll prefer to keep it as is for now.
Moreover I think form is more readable than previous and with less extra apis/helpers involved).

Nokel81 replied:
I completely disagree, if you can show that it wasn't working then sure go ahead. But from all the docs of how JS's `Array.from()` works. The previous code was correct.

ixrock replied:
Yeah, maybe your version is correct. But I prefer mine for the sake of simplicity.

Nokel81 replied:
How is it more simple? Also a migration PR is not the place for this sort of thing.

ixrock replied:
It simple because:
- uses only native JS apis
- don't use generators/iterators (less complexity imho)

Let's ask from @lensapp/lens-maintainers which form they think more simple.
Anyway, I can revert these just to avoid tedious arguing with you.

ixrock replied:
Reverted.

Nokel81 commented on src/common/catalog/catalog-entity-registry.ts:
```diff
@@ -19,27 +19,33 @@
-  @action addComputedSource(id: string, source: IComputedValue<CatalogEntity[]>) {
-    this.sources.set(id, source);
+  @action
+  addObservableSource(id: string, source: IObservableArray<CatalogEntity> | IComputedValue<CatalogEntity[]>) {
```
Again, please revert changes like this. This is not at all related to changing to `mobx6`.

Nokel81 commented on src/extensions/registries/page-registry.ts:
```diff
@@ -34,21 +34,18 @@ export interface PageRegistration {
    */
   id?: string;
-  params?: PageParams<string | ExtensionPageParamInit>;
+  params?: PageParams<string | Omit<PageParamInit<any>, "name" | "prefix">>;
```
These changes are not related to `mobx6`. Please move them to a different PR.

ixrock replied:
It's partially related because `PageParam` helper depends on `mobx-observable-history` and it was refactored to align with new version.

Nokel81 commented on src/main/catalog-sources/kubeconfig-sync.ts:
```diff
@@ -68,7 +74,7 @@ export class KubeconfigSyncManager extends Singleton {
     }
 
-    this.syncListDisposer = UserStore.getInstance().syncKubeconfigEntries.observe(change => {
+    this.syncListDisposer = UserStore.getInstance().syncKubeconfigEntries.observe_(change => {
```
Why the `_`?

ixrock replied:
There is no `.observe()` anymore in `mobx-6`

Nokel81 replied:
Yes it does: https://mobx.js.org/intercept-and-observe.html

Just is a free function instead.

ixrock replied:
Replaced.

Nokel81 commented on src/main/catalog-sources/kubeconfig-sync.ts:
```diff
@@ -248,17 +255,25 @@ async function watchFileChanges(filePath: string): Promise<[IComputedValue<Catal
     disableGlobbing: true,
   });
-  const rootSource = new ExtendedObservableMap<string, ObservableMap<string, RootSourceValue>>(observable.map);
+  const rootSource = new ObservableMap<string, ObservableMap<string, RootSourceValue>>();
```
Please revert unless my class doesn't work for some reason. This change is extraneous.

ixrock replied:
@Nokel81 what the purpose for that extendable map over normal? @jakolehm wdyt?

ixrock replied:
Reverted

Nokel81 commented on src/renderer/api/endpoints/helm-releases.api.ts:
```diff
@@ -155,24 +155,26 @@ export async function rollbackRelease(name: string, namespace: string, revision:
 }
 
-@autobind()
+export interface HelmRelease {
+  appVersion: string;
+  name: string;
+  namespace: string;
+  chart: string;
+  status: string;
+  updated: string;
+  revision: string;
+}
```
This seems strange, I know that it is legal but what happens if you import "HelmRelease" somewhere else, does it mean both?

Should we change this to a different name then?

ixrock replied:
Nope, don't we think we should. Class can be merged with interfaces with same name within a module.
I guess it could be considered it as class type extension, but without fields definition in case of `{useDefineForClassFields: true}`

Nokel81 replied:
What is the benefit of moving this out then? There is `import type { HelmRelease }` already in typescript (see #2777)

ixrock replied:
Consider an example:

```ts
class Test {
  constructor(data: any){
    Object.assign(this, data);
  }
}

class Test2 extends Test {
  public notGood: any;
  
  constructor(data: any){
    super(data);
    console.log(this.notGood); // prints "undefined" when compiled with {useDefineForClassFields: true}
  }
}

console.log(
  new Test2({notGood: true}),
);
```

If we move definition from `Test2` to external interface within class module, then all good and console prints `true`:

```ts
interface Test2 {
 notGood: any;
}
```

Nokel81 replied:
Wouldn't just not using `Object.assign` also fix this problem?

ixrock replied:
Don't use `Object.assign` at all and remember this choice for every inheriting class? Sounds weird.

Nokel81 replied:
Exactly, `Object.assign` is weird and is contrary to how Typescript would like you do it. It prevents detecting things like guaranteed assignment in constructors.

jakolehm replied:
Btw why did we enable `useDefineForClassFields` in this PR? Does Mobx6 require it?

Nokel81 commented on src/renderer/api/endpoints/ingress.api.ts:
```diff
@@ -117,6 +113,17 @@ export class Ingress extends KubeObject {
   };
+}
+
+export class Ingress extends KubeObject {
```
What I meant is that this should be written as:

```typescript
export interface IngressSpec {
}

export interface IngressMetadata extends IKubeMetadata {
  tls: {
    secretName: string;
  }[];
  rules?: {
    host?: string;
    http: {
      paths: {
        path?: string;
        backend: IIngressBackend;
      }[];
    };
  }[];
  // extensions/v1beta1
  backend?: IExtensionsBackend;
  // networking.k8s.io/v1
  defaultBackend?: INetworkingBackend & {
    resource: {
      apiGroup: string;
      kind: string;
      name: string;
    }
  }
}

export interface IngressStatus {
  loadBalancer: {
    ingress: ILoadBalancerIngress[];
  };
}

export class Ingress extends KubeObject<IngressMetadata, IngressStatus, IngressSpec> {
   ...
}
```

ixrock replied:
You can do that in follow-up PRs if you like.

ixrock committed:
clean up / responding to comments

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: app-crash when navigating to catalog from active cluster-view, refactoring `catalog-entity-store`

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
catalog-pusher clean up, replaced .observe_() to external observe() helper from "mobx"

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/jest.setup.ts

ixrock committed:
fix: catalog's items stale/non-observable (after connection to the cluster status still "disconnected"), lint-fixes

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: Catalog is empty after closing main-window and re-opening app from Tray

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: HotBar's icon context menu items non-observable (no "disconnect cluster", etc.)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/common/hotbar-store.ts
#	src/renderer/components/+catalog/catalog.tsx

ixrock committed:
lint-fix/license check

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: redirect to catalog when disconnecting active cluster

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: refresh visibility of active cluster-view on switching from hotbar/catalog

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
updated package.json for built-in extensions to use "*" version for packages served from main app

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/extensions/cluster-feature.ts
#	src/renderer/components/+catalog/catalog-add-button.tsx
#	src/renderer/components/+catalog/catalog.tsx
#	src/renderer/components/cluster-settings/components/cluster-prometheus-setting.tsx
#	src/renderer/components/cluster-settings/components/remove-cluster-button.tsx

ixrock committed:
- added missing makeObservable(this) to metrics-settings.tsx
- updated package-lock.json for built-in extensions
- lint fixes

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	extensions/kube-object-event-status/package-lock.json
#	extensions/metrics-cluster-feature/package-lock.json
#	extensions/node-menu/package-lock.json
#	extensions/node-menu/package.json
#	extensions/pod-menu/package-lock.json
#	extensions/pod-menu/package.json

ixrock committed:
master-merge clean up fix, updated package-lock.json for built-in extensions after `make clean-extensions && make build-extensions`

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix unit-tests

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/common/catalog/catalog-entity-registry.ts
#	src/common/cluster-store.ts
#	src/extensions/registries/page-registry.ts
#	src/main/catalog-pusher.ts
#	src/main/catalog-sources/kubeconfig-sync.ts
#	src/main/extension-filesystem.ts
#	src/renderer/api/endpoints/configmap.api.ts
#	src/renderer/api/endpoints/job.api.ts
#	src/renderer/api/endpoints/resource-quota.api.ts
#	src/renderer/api/endpoints/secret.api.ts
#	src/renderer/api/kube-object.ts
#	src/renderer/api/kube-watch-api.ts
#	src/renderer/components/+apps-releases/release-menu.tsx
#	src/renderer/components/+catalog/catalog-add-button.tsx
#	src/renderer/components/+cluster/cluster-issues.tsx
#	src/renderer/components/+custom-resources/crd-resources.tsx
#	src/renderer/components/+entity-settings/entity-settings.tsx
#	src/renderer/components/+storage-volumes/volume-details-list.tsx
#	src/renderer/components/+user-management-roles-bindings/role-binding-details.tsx
#	src/renderer/components/+workloads-pods/pod-details-list.tsx
#	src/renderer/components/cluster-manager/cluster-view.tsx
#	src/renderer/components/cluster-settings/components/cluster-metrics-setting.tsx
#	src/renderer/components/cluster-settings/components/show-metrics.tsx
#	src/renderer/components/dock/__test__/dock-tabs.test.tsx
#	src/renderer/components/dock/log.store.ts
#	src/renderer/components/dock/logs.tsx
#	src/renderer/components/dock/upgrade-chart.store.ts
#	src/renderer/components/icon/icon.tsx
#	src/renderer/components/kube-object/kube-object-menu.tsx
#	src/renderer/components/notifications/notifications.store.tsx
#	src/renderer/hooks/useStorage.ts
#	src/renderer/navigation/page-param.ts

ixrock committed:
master-merge fixes

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
make lint happy

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/common/catalog/catalog-category-registry.ts
#	src/renderer/api/catalog-entity-registry.ts

ixrock committed:
reverted some changes, removed auto-opening devtools in dev-mode

Signed-off-by: Roman <ixrock@gmail.com>

jakolehm commented on extensions/kube-object-event-status/package.json:
```diff
@@ -18,6 +18,9 @@
   "dependencies": {},
   "devDependencies": {
     "@k8slens/extensions": "file:../../src/extensions/npm/extensions",
+    "react": "latest",
```
We can probably remove these.

ixrock replied:
Yes. I'll create quick follow-up once this merged (since 2 approves already done)

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/main/catalog-pusher.ts
#	src/main/catalog-sources/kubeconfig-sync.ts
#	src/main/catalog/catalog-entity-registry.ts

ixrock committed:
merge fixes

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/main/catalog-pusher.ts
#	src/main/index.ts
#	src/renderer/api/catalog-entity-registry.ts
#	src/renderer/components/+catalog/catalog.tsx

ixrock committed:
master-merge conflict fixes:
- proper handling and navigating into catalog's active category via URL-builder

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/common/catalog/catalog-entity.ts:
```diff
@@ -56,8 +56,13 @@ export abstract class CatalogCategory extends EventEmitter {
+
   public getId(): string {
-    return `${this.spec.group}/${this.spec.names.kind}`;
+    return `${this.spec.group}--${this.spec.names.kind}`;
```
Why does this need to be URL compatible?

ixrock replied:
To navigate to particular category with `catalogURL()`, see: `catalog.route.ts`

ixrock replied:
For example `/` cannot be used, otherwise see exception from the comment above.

Nokel81 replied:
But it was working before in Alex's PR.

Nokel81 commented on src/renderer/components/+catalog/catalog.route.ts:
```diff
@@ -23,11 +23,10 @@ import type { RouteProps } from "react-router";
 }
 export const catalogRoute: RouteProps = {
-  path: "/catalog/:group?/:kind?"
+  path: "/catalog/:categoryId?" // `categoryId` == `${group}--${kind}`
```
What is wrong with the old way?

ixrock replied:
Yeah, probably having splitted url-params for the route is better. Fixed.

Nokel81 commented on extensions/kube-object-event-status/package.json:
```diff
@@ -18,6 +18,9 @@
     "@k8slens/extensions": "file:../../src/extensions/npm/extensions",
+    "react": "latest",
+    "mobx": "latest",
+    "mobx-react": "latest",
```
How were these being supported before?

ixrock replied:
Some leftovers. Removed.

Nokel81 commented on extensions/kube-object-event-status/tsconfig.json:
```diff
@@ -13,6 +13,7 @@
     "esModuleInterop": true,
     "allowSyntheticDefaultImports": true,
     "experimentalDecorators": true,
+    "useDefineForClassFields": true,
```
I thought you were getting rid of this change.

ixrock replied:
Nope. It was returned, see commits history. 
This is actually a part of official mobx-6 migration guide.

Nokel81 replied:
Then I think we should do that in a separate PR before this one gets merged. It is a pretty big change in and of itself.

aleksfront replied:
I guess we should consider to keep this into current PR. The reasons:
1. It's part of migration guide for TS users https://mobx.js.org/migrating-from-4-or-5.html. And since we want to use decorators, we should include this.
2. It's impossible to satisfy TS compiler in current PR without `useDefineForClassFields`, so after merge `master` would be broken?

ixrock replied:
@aleksfront what do you mean by after merge `master` would be broken? 
`useDefineForClassFields: true` required for class fields definitions with decorators without initial value, e.g.

```ts
class Test {
 @observable someVar: string; // no initial value, doesn't work with mobx-6
}
```

Nokel81 reviewed (dismissed):
This PR is pretty big by itself without all the changes related to turning on `useDefineForClassFields`. 

I think that should be split out into its own PR since it is not part of `mobx-6` and we can discuss how to go about it there.

ixrock committed:
reverting splitted params for catalog's page route to "/catalog/:group?/:kind?"

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
clean-up: remove app's injecting dependencies from `extensions/kube-object-event-status/package.json`

Signed-off-by: Roman <ixrock@gmail.com>

ixrock commented on src/renderer/components/+catalog/catalog.tsx:
```diff
@@ -86,17 +88,6 @@ export class Catalog extends React.Component<Props> {
       }),
-
-      // select initial category if nothing yet selected (via url-params)
-      when(() => this.categories.length > 0, () => {
```
Default category for the catalog is `Browse` if no additional params provided via URL. Wdyt? Good/bad/ugly? ©

Nokel81 replied:
I think that is a good idea.

Nokel81 commented on src/common/catalog/catalog-entity.ts:
```diff
@@ -56,13 +56,14 @@ export abstract class CatalogCategory extends EventEmitter {
+    const [group, kind] = id.split("/") ?? [];
+
+    return { group, kind };
   }
```
This could be instead:

```ts
public getRouteParams(): ICatalogViewRouteParam {
  return {
	group: this.spec.group,
    kind: this.spec.names.kind,
  }
}
```

ixrock replied:
It's a`static` method. It's done to match with selected entity: 
See `catalog.tsx`:

```tsx
  onTabChange = (categoryId: string) => {
    const { group, kind } = CatalogCategory.parseId(categoryId);

    if (group && kind) {
      navigate(catalogURL({ params: { group, kind } }));
    } else {
      navigate(catalogURL());
    }
  };
```

Nokel81 replied:
It could also be:

```ts
  onTabChange = (categoryId: string) => {
    const activeCategory = this.categories.find(category => category.getId() === tabId);

    navigate(catalogURL({ params: activeCategory?.getRouteParams() }));
  };
```

ixrock commented on docs/extensions/guides/stores.md:
```diff
@@ -42,18 +42,17 @@ export class ExamplePreferencesStore extends Store.ExtensionStore<ExamplePrefere
+    return {
       enabled: this.enabled
-    }, {
-      recurseEverything: true
```
@stefcameron it's removed from current docs. Is that what you asked? ^

stefcameron replied:
If by "current" you mean the docs associated with Lens 5 changes, then yes. If by "current" you mean Lens 4.2.4 docs, then it's still there (and should be, I guess, since 4.2.4 still users `mobx` < 6).

ixrock replied:
By "current" I mean a version mentioned in project's `package.json` or above whenever it's gonna be released.
Otherwise in current branch all mentioning about `recurseEverything:true` was removed.

stefcameron replied:
Ok, thank you! 👍

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

# Conflicts:
#	src/renderer/components/+extensions/extensions.tsx

ixrock committed:
master-merge fix: added missing makeObservable(this) for extensions.tsx

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into mobx6-migration

ixrock committed:
fix: catalog entity context menu stale/unobservable

Signed-off-by: Roman <ixrock@gmail.com>

jakolehm reviewed (approved):
Awesome job 🏆
